Numbered parser (In Progress)

christinem opened the pull request:
This builds on top of #989.

This PR attempts to create a parser which parses post categories that are listed using a numbered list. Currently it does grab and parse post categories that are listed using a numbered list, but not fully (there are still many bugs), and they are not entered into the database yet.

david-yz-liu commented:
@christinem I'm not sure if you're planning on updating this, but it's good to do a merge from master to clean up the diff. (Also there are compiler warnings.)

christinem commented:
@david-yz-liu I added some more commits to this, and did the merge from master and fixed the warnings. I think this is about it that I'll be adding to this particular PR (unless any fixes are needed of course).

david-yz-liu commented on app/WebParsing/ParsecCombinators.hs:
```diff
@@ -7,8 +7,9 @@ module WebParsing.ParsecCombinators
+     parseNumberedLine,
      text, parseAll) where
-
+        
```
Delete this blank space.

david-yz-liu commented on app/WebParsing/PostParser.hs:
```diff
@@ -5,12 +5,13 @@ import qualified Data.Text as T
 import Control.Monad.Trans (liftIO)
 import Text.HTML.TagSoup
 import Text.HTML.TagSoup.Match
+import Data.List.Split
```
Don't do unqualified imports (specify which function is being imported).

david-yz-liu commented on app/WebParsing/PostParser.hs:
```diff
@@ -20,10 +21,12 @@ addPostToDatabase programElements = do
-        liPartitions = partitions isLiTag requirements
+        liPartitions = map parseLi $ partitions isLiTag requirements
+        numberedPartitions = filter (not . T.null) $ map parseNumberedPartition $ getNumberedPartitions requirements
+        nonEmptyPartitions = if (null liPartitions) then numberedPartitions else liPartitions
```
Don't use parentheses around the if condition.

david-yz-liu commented on app/WebParsing/PostParser.hs:
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
         Left _ -> ""
+
+getNumberedPartitions :: [Tag T.Text] -> [[Tag T.Text]]
+getNumberedPartitions tags = do
```
`do` is unnecessary here.

david-yz-liu commented on app/WebParsing/PostParser.hs:
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
+getNumberedPartitions :: [Tag T.Text] -> [[Tag T.Text]]
+getNumberedPartitions tags = do
+    let pTags = partitions isPTag tags
+        pTagsWithoutBr = map (splitWhen isBRTag) pTags
```
You can simplify this by using `concatMap`.

david-yz-liu commented on app/WebParsing/PostParser.hs:
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
+        pTagsWithoutBr = map (splitWhen isBRTag) pTags
+    concat pTagsWithoutBr
+    where
+        isPTag tag = isTagOpenName "p" tag
```
Don't define separate functions; inline them. Note that `isPTag = isTagOpenName "p"`.

christinem commented:
@david-yz-liu Fixes made.

david-yz-liu commented:
@christinem sorry this took a while, but thanks!
